### PR TITLE
[primTorch] Implement one-dimensional fft transforms

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80737
* #80736
* #80571
* __->__ #80570
* #80182

### CODE REVIEW

File: torch/_prims/__init__.py [hunk 1]
```diff
@@ -2588,3 +2594,132 @@ def _uniform_aten(
...
+    dim = utils.canonicalize_dims(input.ndim, dim)
+    utils.validate_no_repeating_dims(dim)
+
+    shape = list(input.shape)
+    if onesided:
+        last_dim = dim[-1]
+        shape[last_dim] = shape[last_dim] // 2 + 1
+
+    dtype = utils.corresponding_complex_dtype(input.dtype)
+    # TODO: Actual result strides may differ, is that okay?
```
Comment (mruberry) [inline torch/_prims/__init__.py:2614]: It's OK for strides to differ currently, yes

File: torch/_prims/__init__.py [hunk 2]
```diff
@@ -2588,3 +2594,132 @@ def _uniform_aten(
...
+
+
+def _fft_r2c_aten(
+    input: TensorLike,
+    *,
+    dim: DimsSequenceType,
+    onesided: bool,
+):
+    normalization = 0  # No normalization
+    dim = utils.canonicalize_dims(input.ndim, dim)
```
Comment (mruberry) [inline torch/_prims/__init__.py:2626]: dim canonicalization usually happens in references, not primitives, so primitives don't have to concern themselves with these details

in this case a wrapper `refs.fft_r2c` can be added to handle input processing and error checking

File: torch/_prims/utils.py [hunk 3]
```diff
@@ -405,7 +405,17 @@ def canonicalize_dim(rank: int, idx: int) -> int:
 
 # Takes a dimension or sequence of dimensions and "wraps" them,
 # mapping negative offsets to positive ones
-def canonicalize_dims(rank: int, indices: DimsType) -> DimsType:
+@overload
```
Comment (mruberry) [inline torch/_prims/utils.py:408]: Nice typing refinement

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -2931,3 +2931,4 @@ def trace(self: TensorLikeType) -> TensorLikeType:
 
 import torch._refs.nn.functional
 import torch._refs.special
+import torch._refs.fft
```
Comment (mruberry) [inline torch/_refs/__init__.py:2934]: Nice namespace creation

File: torch/_refs/fft.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+NormType = Union[None, Literal["forward"], Literal["backward"], Literal["ortho"]]
+_NORM_VALUES = {None, "forward", "backward", "ortho"}
+
+
+def _apply_norm(
+    x: TensorLike, norm: NormType, signal_numel: int, forward: bool
+) -> TensorLikeType:
+    if norm not in _NORM_VALUES:
+        raise RuntimeError(f"Invalid normalization mode: {norm}")
```
Comment (mruberry) [inline torch/_refs/fft.py:27]: Substitute with `check`

File: torch/_refs/fft.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+    if norm == "ortho":
+        return prims.mul(x, 1 / math.sqrt(signal_numel))
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_refs/fft.py:30]: (`*` is equivalent to `torch.mul`, so this policy can also make these expressions more compact)

File: torch/_refs/fft.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    dtype: torch.dtype, require_complex: bool, device: torch.device
+) -> torch.dtype:
+    if dtype.is_complex:
+        return dtype
+
+    # Promote integral to default float type
+    if not dtype.is_floating_point:
+        dtype = torch.get_default_dtype()
+
+    is_rocm = False  # TODO: How to discern rocm from CUDA?
```
Comment (mruberry) [inline torch/_refs/fft.py:48]: I think we can remove this check. We want trace executors to enable primitive, and if a trace executor (like ROCm) doesn't implement a particular primitive (or primitive x dtype combination, as is the case here), then it should thrown an error when that primitive is called

This may seem odd given that we're suggesting other error checks and input validation happen in the reference implementations and not the primitive implementations, but it's a question of semantic errors vs trace executor errors

File: torch/_refs/fft.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    if dtype == torch.half and (is_rocm or device.type != "cuda"):
+        raise RuntimeError("Unsupported dtype Half")
+
+    if require_complex:
+        dtype = utils.corresponding_complex_dtype(dtype)
+
+    return dtype
+
+
+def _promote_tensor_fft(
```
Comment (mruberry) [inline torch/_refs/fft.py:58]: Comment for each helper (and I'll stop repeating this comment now)

Comment (mruberry) [reply torch/_refs/fft.py:58]: `_maybe_promote_tensor_fft` might be more consistent with our naming elsewhere

File: torch/_refs/fft.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    t: TensorLikeType, require_complex: bool = False
+) -> TensorLikeType:
+    cur_type = t.dtype
+    new_type = _promote_type_fft(cur_type, require_complex, t.device)
+    if cur_type == new_type:
+        return t
+    return prims.convert_element_type(t, new_type)
+
+
+# Fixes the shape of x such that x.size(dims[i]) == sizes[i],
```
Comment (mruberry) [inline torch/_refs/fft.py:68]: Nice comment

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -6357,11 +6357,14 @@ def sample_inputs_spectral_ops(self, device, dtype, requires_grad=False, **kwarg
         # cuFFT supports powers of 2 for half and complex half precision
         # NOTE: For hfft, hfft2, hfftn, irfft, irfft2, irfftn with default args
         # where output_size n=2*(input_size - 1), we make sure that logical fft size is a power of two
-        if self.name in ['fft.hfft', 'fft.irfft']:
+        if self.name in ['fft.hfft', 'fft.irfft',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6360]: nit: This is why I prefer partialing the sample inputs functions with kwargs that control its behavior vs. matching the actual name exactly, but nbd

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -20077,6 +20084,30 @@ def __init__(
 
         super(ElementwiseBinaryPythonRefInfo, self).__init__(**ukwargs)
 
+class SpectralFuncPythonRefInfo(SpectralFuncInfo):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:20087]: Nice addition

**Review by mruberry — COMMENTED** [review]
This looks great @peterbell10! I'm curious to hear your thoughts on working with primitives and references, too.

I made a few mostly minor comments inline for your review, principally just style/best practices updates

**Comment by peterbell10** [comment]
@mruberry my experience was that writing refs mostly entailed translating code directly from C++ to python, with only some minor tweaks. Though it certainly helped that I wrote the operators originally with a very similar composite-primitive split style. 

Some of the test failures I got while working on it were very opaque though. Particularly those coming from the executor tests, where error messages were coming from asserts deep in the implementation of `torch.fx`. I would struggle to re-create them to open an issue though.

**Review by mruberry — APPROVED** [review]
Stamped!
